README
* 2023-01-28 *

This is openzz 1.0.4-5, a fork from the catseye github repository.
I have managed to fix the build for MacOs.

The code is very old and had a lot of assumption that caused segfaults.
Most of the time the problem is that pointers (64bit on Mac) are passed as int (32bit).

I have also fixed the test suite that now pass entirely, including the dynamic linking.

To build the program and the libraries you have to cd to the root directory of the distribution and then run:

./configure
make all

To test:

./test.sh

or

cd testsuite
make zztest

To install the program and the libraries:

make install


Matteo

------------------

This is OpenZz 1.0.4-4ce1, a fork of the original OpenZz distribution
based on version 1.0.4-4.  The original README can be found below, following
the first line of equals signs in this file.

This fork was made because development on OpenZz was stalled, I have a
project which requires it ( http://catseye.tc/node/Zzrk.html ), and I
needed to make changes to the sources to get them to build on the latest
version of gcc.

Shortly thereafter, upgrades to (apparently) libtool(?) caused automake
to generate Makefiles which went into infinite loops at the shell level
(`pstree` showed what looked like an infinite tree of `sed` and `bash`)
which would take down my laptop quite readily.

So, I added a simple script `build.sh` to build the `ozz` executable
directly, without messing around with libraries or anything.  It works
on Ubuntu and NetBSD, and will (I'm sure) work on other flavours of Linux
and other BSDs with only minor modifications.

Running `build.sh` is recommended.  Run `autogen.sh` and `configure` at
your own risk.

Also note: several of the tests now fail.  The core semantics seem alright,
so these tests have been skipped.  The specific failures are:
    
*   deref_param.zz: failure of the test suite; it shouldn't be caring about
    an internal address, which isn't stable from run to run
*   double.zz: using floating-point values segfaults
*   list_in_cond.zz: the syntax "a.b" seems to be a syntax error
*   numerictypes.zz: 64-bit values wrap to 32 bits on a 32-bit archiecture
*   tagdtor.xx: can't open the specified .so file

The verdict is that this version of OpenZz does not support:

*   floating-point values
*   64-bit integer values
*   the "a.b" syntax (whatever that means exactly)
*   loading ".so" files

Which is fine by me so long as it can still play Zzrk.

-Chris

===============================================================================
                 Zz - Dynamic Lexical Parser README File
===============================================================================

Zz is a dynamic parser which is currently being developed as a front end to
gcc to implement compilation of new languages (ie. TAO and others), for APE
systems.  These are custom build parallel processing computers used for LQCD
physics research.

See:
  http://chimera.roma1.infn.it/ape.html

UPDATES
 See the NEWS and ChangeLog files for information on the latest changes
 Some support is being formulated at the project directory on SourceForge:
  http://openzz.sourceforge.net  

LICENSE
 The Zz library is released under the GNU Lesser GPL (LGPL).
 Parts of Zz are released under the GPL - see the "COPYING" file in the
   project root.

DOCUMENTATION
 See the "doc" subdirectory

APE
 Whence the name "APE"?  "Ape" is the Italian word for "bee", the mascot of
 the APE project. (INFN, the funder of the project is based in Italy and the
 software is primarily developed there).


ChangeLog
2023-02-01  Matteo Fago <matteo.fago@gmail.com>
	* Some files needed by configure script where missing

2023-01-29  Matteo Fago <matteo.fago@gmail.com>
	* bump version to to 1.0.4-5

	* src/*.c:
		Fixed to compile on 64 bit platforms
		Fixed obsolete syntax for function declaration
		Removed compile warnings
	* testsuite
		Fixed linking of dynamic library from zz for MacOs (non portable!)
		Fixed zz_test.sh for new shells

2002-06-03  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* openzz.spec: bumped to 1.0.4-4
	
	* src/rule.c, src/scope.c, testsuite/when_delete_scope.zz:
	activated 'when delete scope' syntax and added a test case.

	* src/Makefile.am: moved zzi.c to libozzi for interactive
	support. bumped lib version.

	* src/dumpnet.c: some indentations.

	* err.c, zz.h: exported total_error_n as zz_get_error_number().

	* kernel.c: small comment.

	* param.c, zz.h: exported param_level as
	zz_get_param_stack_depth().
	
2002-05-09  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* src/Makefile.am:
	moved version-info for 1.0.4.

	* src/avl.c:
	worked around a bug in node cached allocator for gcc 2.96-98.

	* src/param.c, src/source.c:
	indenting & makeup.

	* src/sys.c, src/zlex.h, src/kernel.c:
	moved some macros. fixed s_load_lib and changed ABI.

	* src/zzdoc_dynamiclibs.xml:
	documented above change.
	
	* src/util.c, src/zz.h:
	added zz_scnt_get_tag_name().

	* src/zz.c:
	-h dumps usage.

	* testsuite/Makefile.am, testsuite/tagdtor.c, testsuite/tagdtor.zz:
	a new test case for dynamic loading of library which adds a new
	type (tag + ctor). tested param ctor & dtor calling.
	
2002-04-09  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* testsuite/Makefile.am:
	added povoidrule.zz.

	* testsuite/zz_test.sh:
	fixed as of zz->ozz.

	* testsuite/return.zz, testsuite/utilities.zz:
	fixed as now real defaults to float.
	
2002-03-28  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* testsuite/popvoidrule.zz:
	added popvoidrule bug test case to CVS.

2002-03-09  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* doc/RPM-howto:
	contributed by zuliani@nergal.it.
	
	* openzz.spec, configure.in, src/Makefile.am: 
	bumped 1.0.2 release number. bumped library version number. added
	table.h.

	* src/zz.c:
	added stuff to usage infos.
	
2002-03-09  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* openzz.spec:
	contributed by zuliani@nergal.it. adapted to changes in 1.0.1.

	* src/Makefile.am:
	moved to ozz prefix.

2002-02-13  Davide Rossetti <davide.rossetti@roma1.infn.it>

	* configure.in:	
	bumped 1.0.1 release number. fixed proper readline handling on
	RH7.2 via libtermcap dependency (thx to Francesco Zuliani
	<zuliani@nergal.it>).
	
	* src/zz.h, src/action.c:
	added zz_bind_get_ret_value() needed by TAO100 compiler.
	WARNING zz_parse_tt() is not really present in libzz!!
	
	* AUTHORS:
	fixed authors email addrs.

2002-02-12  Davide Rossetti <davide.rossetti@roma1.infn.it>
	
	* src/zz.h, src/zlex.c, src/zlex.h:
	add more params to zz_lex_add_new_tag (ctor,dtor,cast) breaking 
	binary compatibity.
	let zlex_realconst_double = 0 so now real consts default to float type.
	some makeup.
	
	* src/Makefile.am:
	bumped major version number.

	* src/sys.c:
	debug zz_trace in s_add() waiting for tag_of by James.

	* src/zz.h, src/zzbind.h:
	made C++ friendly.

2002-02-07  Davide Rossetti <davide.rossetti@roma1.infn.it>
	
	* src/zz.c:
	removed clumsy zz param syntax; changed startup info; default is not verbose
	
2002-01-11  James Brooks <james.brooks@roma1.infn.it>

	* src/sys.c
	Altered "get_rank_for_type()" to take params by reference instead of by value (incurring copy)

2002-01-08  James Brooks <james.brooks@roma1.infn.it>

	* src/zlex.c:
	Bug fix for scientific notation: 3.3e+2 was returning as 5.3

	* testsuite/zz_test.sh
	Added new test files and made test script more userfriedly - accepts
	command line params.

2002-01-07  James Brooks <james.brooks@roma1.infn.it>

	* src/kernel.c, src/sys.c:
	Added "tag_of(param)" utility function to determine tag type of an identifier
	
2001-12-18  James Brooks  <james.brooks@roma1.infn.it>

	* testsuite/boolean.zz:
	Boolean comparisson operators (>=, <=) weren't setting tag type to int - caused bugs.
	Inserted test case for >= and <= operaetors - caught and fixed related bug.

	* src/list.c, src/sys.c:
	Boolean comparisson operators (>=, <=) weren't setting tag type to int - caused bugs.

2001-12-17  dicarlo  <federico.dicarlo@roma1.infn.it>

	* src/util.c: included here zz_parse_string()

2001-12-17  James Brooks  <james.brooks@roma1.infn.it>

	* src/Makefile.am, src/Makefile.orig, src/sys.c, src/util.c:
	New file for keeping utility functions in one place

2001-12-14  James Brooks  <james.brooks@roma1.infn.it>

	* src/parse.c: Typo

	* src/source.c, src/sys.c:
	Moved "zz_parse_string(s)" function from source.c to sys.c

	* Makefile.am, configure.in: Still adjusting...

	* autogen.sh:
	Need aclocal call - without it necessary macros are missing for autoconf

	* src/Makefile.am:
	Still working on this - think we have a rough working draft though
	Work in progress

	* Makefile.am, autogen.sh, configure.in:
	Still working on this - think we have a rough working draft though

2001-12-14  dicarlo  <federico.dicarlo@roma1.infn.it>

	* src/kernel.c, src/source.c, src/sys.c, src/zzi.c: Added in kernel:
	/stat -> /include <>
	/stat -> /add_includedir
	/stat -> /print_includedirs

	moved zz_parse_string from zzi.c to source.c

2001-12-13  Davide Rossetti  <davide.rossetti@roma1.infn.it>

	* src/zz.h: added

	* configure.in:
	fixed spaces after \ and AC_INIT wants an existing filename

2001-12-13  James Brooks  <james.brooks@roma1.infn.it>

	* src/printz.c, src/printz.cc:
	Renamed(and converted) printz.cc to printz.c

	* src/Makefile.am, src/Makefile.orig: Converting ".cc" files to ".c"

2001-12-13  Davide Rossetti  <davide.rossetti@roma1.infn.it>

	* ChangeLog: updated

	* src/kernel.c:
	some comments. added star rule for indirect param definition


src/zz.h
/* 
    Zz Dynamic Parser Library
    Copyright (C) 1989 - I.N.F.N - S.Cabasino, P.S.Paolucci, G.M.Todesco

    The Zz Library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    The Zz Library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/*
  Zz library exported functions

  $Id: zz.h,v 1.9 2002/06/03 11:06:13 kibun Exp $
 */

#ifndef __ZZ_H__
#define __ZZ_H__

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifdef SIZEOF_VOID_P
#define _ZZ_RET_SIZE_ SIZEOF_VOID_P
#else
// If there is no configure script to help us we assume we are on a 64 bit platform
#define _ZZ_RET_SIZE_ 8
#endif

#if _ZZ_RET_SIZE_ == SIZEOF_INT

typedef int zz_ret;
#warning "typedef zz_ret as int"

#elif _ZZ_RET_SIZE_ == SIZEOF_LONG

typedef long int zz_ret;
#warning "typedef zz_ret as long"

#elif _ZZ_RET_SIZE_ == SIZEOF_LONG_LONG

typedef long long int zz_ret;
#warning "typedef zz_ret as long long"

#elif _ZZ_RET_SIZE_ == 8

// Default case
typedef u_int64_t zz_ret;
#warning "typedef zz_ret as u_int64_t"

#elif _ZZ_RET_SIZE_ == 4

typedef u_int32_t zz_ret;
#warning "typedef zz_ret as u_int32_t"

#endif

//#include "zlex.h"
//#include "trace.h"

#include <sys/types.h> /* for int32_t, ... */
#include <stdio.h> /* for FILE ... */

#ifdef  __cplusplus
extern "C" {
#endif

const char* zz_version(void);

void zz_init(void);

/* "kernel" scope open/close */
void zz_set_kernel_flag(int bool_value);

void zz_set_output(const char* output_filename);
void zz_set_output_stream(FILE* output_stream);

int  zz_parse_string(const char *s);
int  zz_parse_file(const char* input_source_filename);
int  zz_parse_pipe(void);
  /* this one is really in libozzi.xx */
int  zz_parse_tt(void); 

void zz_set_default_extension(const char* include_file_def_extension);
void zz_set_default_include_dir(const char* default_include_dir);

void zz_set_prompt(const char *prompt);
const char* zz_get_prompt(void);

/* error reporting related stuff */
int zz_get_param_stack_depth();
int zz_get_error_number();

/* scanner extension */
struct s_content;
struct s_tag;

/*   struct s_content accessors */
long    zz_scnt_get_lvalue (struct s_content *);
int32_t zz_scnt_get_ivalue (struct s_content *);
float   zz_scnt_get_fvalue (struct s_content *);
void*   zz_scnt_get_pvalue (struct s_content *);
char*   zz_scnt_get_svalue (struct s_content *);
double  zz_scnt_get_dvalue (struct s_content *);
int64_t zz_scnt_get_llvalue(struct s_content *);

long    zz_scnt_getv_lvalue (struct s_content *, int);
int32_t zz_scnt_getv_ivalue (struct s_content *, int);
float   zz_scnt_getv_fvalue (struct s_content *, int);
void*   zz_scnt_getv_pvalue (struct s_content *, int);
char*   zz_scnt_getv_svalue (struct s_content *, int);
double  zz_scnt_getv_dvalue (struct s_content *, int);
int64_t zz_scnt_getv_llvalue(struct s_content *, int);

void zz_scnt_set_lvalue (struct s_content *, long   );
void zz_scnt_set_ivalue (struct s_content *, int32_t);
void zz_scnt_set_fvalue (struct s_content *, float  );
void zz_scnt_set_pvalue (struct s_content *, void*  );
void zz_scnt_set_svalue (struct s_content *, char*  );
void zz_scnt_set_dvalue (struct s_content *, double );
void zz_scnt_set_llvalue(struct s_content *, int64_t);

const char* zz_scnt_get_tag_name(struct s_content *);

typedef int (*zz_tag_sprint_cb)(char *, struct s_content *);
typedef int (*zz_tag_fprint_cb)(FILE *, struct s_content *);
typedef int (*zz_tag_cdtor)(struct s_content *cnt, const char *param_name);
typedef struct s_content * (*zz_tag_cast)(struct s_content* src, struct s_tag* tgt_type_tag, struct s_content* tgt);

int zz_lex_add_new_tag(const char* tag_name, 
		       zz_tag_sprint_cb, 
		       zz_tag_fprint_cb, 
		       zz_tag_cdtor pctor, 
		       zz_tag_cdtor pdtor, 
		       zz_tag_cast cst);

#define zz_lex_add_new_tag2(TAG_NAME, TAG_SP, TAG_FP) zz_lex_add_new_tag(TAG_NAME, TAG_SP, TAG_FP, 0, 0)

int zz_lex_remove_tag(const char* tag_name);

/* tracing stuff */

#define TRACE_REDUCE   0x1
#define TRACE_ZZACTION 0x2
#define TRACE_SCOPE    0x4
#define TRACE_LRSTACK  0x8
#define TRACE_ALL (TRACE_REDUCE|TRACE_ZZACTION|TRACE_SCOPE|TRACE_LRSTACK)

void zz_set_trace_mask(int mask);

#ifdef  __cplusplus
}
#endif

#endif /* __ZZ_H__ */
